Core/include/Acts/Detector/detail/KdtSurfacesProvider.hpp

```diff
@@ -0,0 +1,177 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2023 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Definitions/Algebra.hpp"
+#include "Acts/Detector/GridAxisGenerators.hpp"
+#include "Acts/Detector/IndexedGridFiller.hpp"
+#include "Acts/Detector/IndexedSurfacesGenerator.hpp"
+#include "Acts/Detector/SupportBuilder.hpp"
+#include "Acts/Surfaces/Surface.hpp"
+#include "Acts/Utilities/BinningData.hpp"
+#include "Acts/Utilities/Helpers.hpp"
+#include "Acts/Utilities/KDTree.hpp"
+
+#include <array>
+#include <stdexcept>
+#include <tuple>
+#include <vector>
+
+namespace Acts {
+
+namespace Experimental {
+
+namespace detail {
+
+/// @brief A wrapper class around a KDTree of surfaces
+///
+/// It also deals with the conversion from global query to
+/// KDTree lookup positions
+///
+template <size_t kDIM = 2u, size_t bSize = 100u,
+          typename reference_generator = PolyhedronReferenceGenerator>
+class KdtSurfaces {
+ public:
+  /// Broadcast the surface KDT type
+  using KDTS =
+      KDTree<kDIM, std::shared_ptr<Surface>, ActsScalar, std::array, bSize>;
+
+  /// Broadcast the query definition
+  using Query = std::array<ActsScalar, kDIM>;
+
+  /// Broadcast the entry
+  using Entry = std::pair<Query, std::shared_ptr<Surface>>;
+
+  /// Constructor from a vector of surfaces
+  ///
+  /// @param gctx the geometry context of this call
+  /// @param surfaces the surfaces to be filled into the tree
+  /// @param casts the cast list from global position into kdtree local
+  /// @param rgen the reference point generator
+  ///
+  KdtSurfaces(const GeometryContext& gctx,
+              const std::vector<std::shared_ptr<Surface>>& surfaces,
+              const std::array<BinningValue, kDIM>& casts,
+              const reference_generator& rgen = PolyhedronReferenceGenerator{})
+      : m_kdt(nullptr), m_casts(casts), m_rGenerator(rgen) {
+    // Simple check if the dimension is correct
+    if (kDIM == 0u or kDIM != m_casts.size()) {
+      throw std::invalid_argument(
+          "KdtSurfaces: dimension and/or cast rules are incorrect.");
+    }
```

Comment:
can this be a static assert?

also isn't `kDIM != m_casts.size()` guaranteed to be `false`?

Reply:
Correct, fixed.